### ISSUE TITLE
Relax the restriction that list literals can only contain elements of type Type.

### ISSUE THREAD

**Issue by ocharles** [dhall-lang/dhall-lang#1120]
The [type inference rules for list literals](https://github.com/dhall-lang/dhall-lang/blob/master/standard/type-inference.md#list) states

> Note that the above rules forbid List elements that are Types. More generally, if the element type is not a Type then that is a type error.

Is there a good reason to have this restriction? What is the problem of changing

```
Γ ⊢ t : T₀   Γ ⊢ T₀ : Type   Γ ⊢ [ ts… ] : List T₁   T₀ ≡ T₁
────────────────────────────────────────────────────────────
Γ ⊢ [ t, ts… ] : List T₀
```

to

```
Γ ⊢ t : T₀   Γ ⊢ [ ts… ] : List T₁   T₀ ≡ T₁
────────────────────────────────────────────────────────────
Γ ⊢ [ t, ts… ] : List T₀
```

?

**Comment by ocharles** [comment]
I suppose one follow-on complication is:

> List is a function from a Type to another Type:

```
──────────────────────
Γ ⊢ List : Type → Type
```

Would no longer be true. Perhaps at this point we're stuck.

**Comment by Gabriel439** [comment]
@ocharles: Yeah, the only way this would work is if `List` became a keyword that had to be saturated with the element type instead of a built-in

**Comment by ocharles** [comment]
Well, that, or implicit arguments to specify the type of the type being applied to the list

**Comment by MonoidMusician** [comment]
what we need is ✨ _universe polymorphism_ ✨

**Comment by winitzki** [comment]
Currently in Dhall it is possible to implement type-level lists   if the kind is monomorphic. For example, one can define the type `ListType` containing elements `t1 : Type`, `t2 : Type`, making an analog of `[ Natural, Bool ]` possible 

Separately we can define the type `ListType2` containing `t1 : Type → Type`,  `t2 : Type → Type`, etc., making an analog of `[ List, Option ]` possible. 

But I'm not sure we can define `ListT k` so that we could have an arbitrary kind `k`, to  avoid having to define `ListType` separately from `ListType2` and from all other similar lists.  

If this is not possible then it's not clear whether it's safe to add such a magical `ListT` to the language as a built-in symbol.
